dlt/transformations/transformation.py

```diff
@@ -1,52 +1,80 @@
 from functools import wraps
-import inspect
-from typing import Callable, Any, Optional, Type, Iterator, List
-
+from typing import Callable, Any, Optional, Type, Iterator, List, cast
 
 import dlt
+import sqlglot
 
 from dlt.common.configuration.inject import get_fun_last_config, get_fun_spec
-from dlt.common.reflection.inspect import isgeneratorfunction
 from dlt.common.typing import TDataItems, TTableHintTemplate
-from dlt.common import logger
-
-from dlt.destinations.dataset.relation import BaseReadableDBAPIRelation
-from dlt.extract.hints import SqlModel
-from dlt.extract.incremental import Incremental
+from dlt.common import logger, json
 
-from dlt.transformations.typing import (
-    TTransformationFunParams,
-)
+from dlt.extract import DltResource
+from dlt.transformations.typing import TTransformationFunParams
 from dlt.transformations.exceptions import (
     TransformationException,
-    UnknownColumnTypesException,
-    TransformationInvalidReturnTypeException,
     IncompatibleDatasetsException,
 )
+from dlt.destinations.dataset.dataset import ReadableDBAPIDataset
+from dlt.common.typing import TDataItem
+from dlt.common.schema.typing import TTableSchema
+
+from dlt.common.exceptions import MissingDependencyException
 from dlt.pipeline.exceptions import PipelineConfigMissing
-from dlt.destinations.dataset import ReadableDBAPIDataset
-from dlt.common.schema.typing import TTableSchemaColumns
-from dlt.extract.hints import make_hints
-from dlt.common.destination.dataset import SupportsReadableRelation
-from dlt.extract import DltResource
-from dlt.transformations.configuration import TransformationConfiguration
-from dlt.common.utils import get_callable_name
 from dlt.common.schema.typing import (
+    TTableSchemaColumns,
     TWriteDisposition,
     TColumnNames,
     TSchemaContract,
     TTableFormat,
     TTableReferenceParam,
 )
-from dlt.extract.exceptions import (
-    CurrentSourceNotAvailable,
-)
+from dlt.transformations.configuration import TransformationConfiguration
+from dlt.common.utils import get_callable_name
+from dlt.extract.exceptions import CurrentSourceNotAvailable
+from dlt.extract.pipe_iterator import DataItemWithMeta
+from dlt.extract.hints import DLT_HINTS_METADATA_KEY, TResourceHints
+from dlt.destinations.dataset.relation import ReadableDBAPIRelation
+
+try:
+    from dlt.helpers.ibis import Expr as IbisExpr
+except (ImportError, MissingDependencyException):
+    IbisExpr = None
+
+try:
+    from dlt.common.libs.pyarrow import pyarrow
+except (ImportError, MissingDependencyException):
+    pyarrow = None
 
 
 class DltTransformationResource(DltResource):
     def __init__(self, *args: Any, **kwds: Any) -> None:
         super().__init__(*args, **kwds)
 
+    def compute_table_schema(self, item: TDataItem = None, meta: Any = None) -> TTableSchema:
```

Comment:
this is good. but we have one more complication. `extract` is caching static tables (where schema will not mutate). we check this like:
```py
def _get_static_table_name(self, resource: DltResource, meta: Any) -> Optional[str]:
        if resource._table_name_hint_fun:
```
maybe we should make this interface explicit? `has_static_hints() -> bool` and for transformations it is always false.

also this should be possibly fixed:
```py
def _get_dynamic_table_name(self, resource: DltResource, item: TDataItem) -> str:
        return self._normalize_table_identifier(resource._table_name_hint_fun(item))
```
so we can hide `._table_name_hint_fun` completely

Reply:
I did this, but we still need `_table_name_hint_fun` in the extractor to get the table name. I also tried to set the has_dynamic_hints to true if there are other dynamic hints, this works but it would change the extractor behavior. I am not sure if there maybe is a bug in there from before? It will only start writing to dynamic tables and updating hints if the table name is dynamic, this does not seem right..

Reply:
this is correct - table name must be dynamic for other hints to be dynamic. but I can't find where it was enforced...

dynamic hints can be pretty slow so the idea was:
- if only table name is dynamic we can still cache hints per table name
- if anything more is dynamic then we must always (for each item) recompute the hints

I'll take look at the code

Reply:
this background why we do the schema caching and when